src/Features/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,572 @@
+                return new JsonLiteralNode(token);
+
+            // Look for constructors (a json.net extension).  We'll report them as an error
+            // in strict model.
+            if (Matches(token, "new"))
+                return ParseConstructor(token);
+
+            // Check for certain literal values.  Some of these (like NaN) are json.net only.
+            // We'll check for these later in the strict-mode pass.
+            Debug.Assert(token.VirtualChars.Length > 0);
+            if (TryMatch(token, "NaN", JsonKind.NaNLiteralToken, out var newKind) ||
+                TryMatch(token, "true", JsonKind.TrueLiteralToken, out newKind) ||
+                TryMatch(token, "null", JsonKind.NullLiteralToken, out newKind) ||
+                TryMatch(token, "false", JsonKind.FalseLiteralToken, out newKind) ||
+                TryMatch(token, "Infinity", JsonKind.InfinityLiteralToken, out newKind) ||
+                TryMatch(token, "undefined", JsonKind.UndefinedLiteralToken, out newKind))
+            {
+                return new JsonLiteralNode(token.With(kind: newKind));
+            }
+
+            if (Matches(token, "-Infinity"))
+            {
+                SplitLiteral(token, out var minusToken, out var newLiteralToken);
+
+                return new JsonNegativeLiteralNode(
+                    minusToken, newLiteralToken.With(kind: JsonKind.InfinityLiteralToken));
+            }
+
+            var firstChar = token.VirtualChars[0];
+            if (firstChar == '-' || firstChar == '.' || IsDigit(firstChar))
+                return ParseNumber(token);
+
+            return new JsonTextNode(
+                token.With(kind: JsonKind.TextToken).AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    string.Format(FeaturesResources._0_unexpected, firstChar.ToString()),
+                    firstChar.Span)));
+        }
+
+        private JsonConstructorNode ParseConstructor(JsonToken token)
+        {
+            var savedInConstructor = _inConstructor;
+            _inConstructor = true;
+
+            var result = new JsonConstructorNode(
+                token.With(kind: JsonKind.NewKeyword),
+                ConsumeToken(JsonKind.TextToken, FeaturesResources.Name_expected),
+                ConsumeToken(JsonKind.OpenParenToken, s_openParenExpected),
+                ParseSequence(),
+                ConsumeToken(JsonKind.CloseParenToken, s_closeParenExpected));
+
+            _inConstructor = savedInConstructor;
+            return result;
+        }
+
+        private static bool TryMatch(JsonToken token, string val, JsonKind kind, out JsonKind newKind)
+        {
+            if (Matches(token, val))
+            {
+                newKind = kind;
+                return true;
+            }
+
+            newKind = default;
+            return false;
+        }
+
+        private static bool Matches(JsonToken token, string val)
+        {
+            var chars = token.VirtualChars;
```

Comment:
I wonder if this should be an extension/method on `VirtualChars`. I'm pretty sure I had to do similar things for stack parsing.

Reply:
if you link me, i can check!